lib/changesreader.js

```diff
@@ -0,0 +1,280 @@
+const EventEmitter = require('events').EventEmitter
+const stream = require('stream')
+const EVENT_BATCH = 'batch'
+const EVENT_CHANGE = 'change'
+const EVENT_SEQ = 'seq'
+const EVENT_ERROR = 'error'
+
+// streaming line breaker
+const liner = () => {
+  const liner = new stream.Transform({ objectMode: true })
+
+  liner._transform = function (chunk, encoding, done) {
+    let data = chunk.toString('utf8')
+    if (this._lastLineData) {
+      data = this._lastLineData + data
+      this._lastLineData = null
+    }
+
+    const lines = data.split(/\s*\n/)
+    this._lastLineData = lines.splice(lines.length - 1, 1)[0]
+    lines.forEach(this.push.bind(this))
+    done()
+  }
+
+  liner._flush = function (done) {
+    this.push(this._lastLineData)
+    this._lastLineData = null
+    done()
+  }
+
+  return liner
+}
+
+// streaming change processor
+const changeProcessor = (ee, batchSize) => {
+  const changeProcessor = new stream.Transform({ objectMode: true })
+  const buffer = []
+  changeProcessor.lastSeq = '0'
+
+  changeProcessor._transform = function (chunk, encoding, done) {
+    // remove last char from string
+    if (chunk[chunk.length - 1] === ',') {
+      chunk = chunk.slice(0, -1)
+    }
+
+    try {
+      const j = JSON.parse(chunk)
+      buffer.push(j)
+      if (buffer.length >= batchSize) {
+        ee.emit(EVENT_BATCH, buffer.splice(0, batchSize))
+      }
+      done()
+    } catch (e) {
+      // look for last_seq
+      const match = chunk.match(/"last_seq":"([^"]+)"/)
+      if (match) {
+        changeProcessor.lastSeq = match[1]
+      }
+      done()
```

Comment:
oh btw, what happens if `match` is false? Should the call be `done(err)`??

Reply:
Don't want to surface an error when paring `}` at the end of a changes feed ;)

Reply:
ahh, that makes sense!

Reply:
hmm, I think the try/catch is handling too many scenarios, and given the `done()` swallow the error it potentially can hide bugs.

Ideally the try catch will only happen if `JSON.parse` fails, and then the catch will try to match using the regexp.. if that works, then `done()` is a correct call, in any other case then `done(err)` should be called.

Please let me know if my analysis is correct, I might be missing some detail about how this processor will be pipe in the chain, from liner -> change

Reply:
ohhh this will produce wrong results if for some reason the `chunk` ends right in between the value of `last_seq`.

Suppose these chunks:

```js
const chunks = [
  '{ "first": 1,',
  ' "second": ',
  ' 2, "third',
  '": 3, "last_seq": 123',
  '456 }',
]

const obj = JSON.parse(chunks.join(''))

console.log(obj.last_seq) /// 123456
```

I feel this is extremely unlucky but totally possible, and given enough time it could very certainly match such scenario.

What do you think?